File: Backend/DikkeTennisLijst.Core/Calculations/ScoreCalculations.cs
using DikkeTennisLijst.Core.Entities;
using DikkeTennisLijst.Core.Shared.Constants;
using DikkeTennisLijst.Core.Shared.Enums;
using Bonus = DikkeTennisLijst.Core.Shared.Constants.BonusConstants;

namespace DikkeTennisLijst.Core.Calculations
{
    public static class ScoreCalculations
    {
        /// <summary>
        /// Calculates the expected score, based on the players elo's and the match format.
        /// </summary>
        /// <param name="player">The elo of the player for which to calculate the expected score</param>
        /// <param name="opponent">The elo of the opponent.</param>
        /// <param name="format"></param>
        /// <returns></returns>
        public static double CalculateExpectedScore(int playerElo, int opponentElo, MatchFormat format)
        {
            var expectedScore = CalculateExpectedScorePerSet(playerElo, opponentElo);

            return format switch
            {
                MatchFormat.BestOf1 => expectedScore,
                MatchFormat.BestOf3 => BestOf3(expectedScore),
                MatchFormat.BestOf5 => BestOf5(expectedScore),
                MatchFormat.BestOf7 => BestOf7(expectedScore),
                _ => throw new ArgumentOutOfRangeException(nameof(format)),
            };
        }

        /// <summary>
        /// Calculates the expected score, based on the players elo's.
        /// </summary>
        /// <param name="player">The elo of the player for which to calculate the expected score</param>
        /// <param name="opponent">The elo of the opponent.</param>
        /// <param name="format"></param>
        /// <returns></returns>
        private static double CalculateExpectedScorePerSet(int playerElo, int opponentElo)
        {
            return 1 / (double)(1 + (double)Math.Pow(10, ((double)opponentElo - playerElo) / EloConstants.Diff));
        }

        /// <summary>
        /// Calculates the winning chance for a best of 3, when the chance to win a set is provided.
        /// q = 1-p;
        /// x = p*p*(p + 3*q);
        /// </summary>
        /// <param name="expectedScore">The chance to win a single set</param>
        /// <returns></returns>
        private static double BestOf3(double expectedScore)
        {
            double E = expectedScore;
            double I = 1 - expectedScore;
            return Math.Pow(E, 2) * (E + (3 * I));
        }

        /// <summary>
        /// Calculates the winning chance for a best of 5, when the chance to win a set is provided.
        /// q = 1-p;
        /// x = p*p*p*(p*p + 5*p*q + 10*q*q);
        /// </summary>
        /// <param name="expectedScore">The chance to win a single set</param>
        /// <returns></returns>
        private static double BestOf5(double expectedScore)
        {
            double E = expectedScore;
            double I = 1 - expectedScore;
            return Math.Pow(E, 3) * (Math.Pow(E, 2) + (5 * E * I) + (10 * Math.Pow(I, 2)));
        }

        /// <summary>
        /// Calculates the winning chance for a best of 7, when the chance to win a set is provided.
        /// q = 1-p;
        /// x = p*p*p*p*(p*p*p + 7*p*p*q + 21*p*q*q + 35*q*q*q);
        /// </summary>
        /// <param name="expectedScore">The chance to win a single set</param>
        /// <returns></returns>
        private static double BestOf7(double expectedScore)
        {
            double E = expectedScore;
            double I = 1 - expectedScore;
            return Math.Pow(E, 4) * (Math.Pow(E, 3) + (7 * Math.Pow(E, 2) * I) + (21 * E * Math.Pow(I, 2)) + (35 * Math.Pow(I, 3)));
        }

        public static (double PlayerOne, double PlayerTwo) CalculateActualScore(Match match)
        {
            var (scorePlayerOne, scorePlayerTwo) = CalculateScoreTotals(match);
            return CalculateActualScore(scorePlayerOne, scorePlayerTwo);
        }

        private static (int PlayerOne, int PlayerTwo) CalculateScoreTotals(Match match)
        {
            int scorePlayerOne = 0;
            int scorePlayerTwo = 0;

            foreach (var set in match.Sets)
            {
                if (set.GamesP1.HasValue && set.GamesP2.HasValue)
                {
                    scorePlayerOne += set.GamesP1.Value * Bonus.Game;
                    scorePlayerTwo += set.GamesP2.Value * Bonus.Game;

                    scorePlayerOne += set.GamesP1 > set.GamesP2 ? Bonus.Set : 0;
                    scorePlayerTwo += set.GamesP2 > set.GamesP1 ? Bonus.Set : 0;

                    scorePlayerOne += set.IsTieBreaker && set.GamesP1 > set.GamesP2 ? Bonus.TieBreaker : 0;
                    scorePlayerTwo += set.IsTieBreaker && set.GamesP2 > set.GamesP1 ? Bonus.TieBreaker : 0;
                }

                scorePlayerOne += set.IsSuperTieBreaker && set.PointsP1 > set.PointsP2 ? Bonus.SuperTieBreaker : 0;
                scorePlayerTwo += set.IsSuperTieBreaker && set.PointsP2 > set.PointsP1 ? Bonus.SuperTieBreaker : 0;
            }

            scorePlayerOne += match.Winner == MatchWinner.PlayerOne ? Bonus.Match : 0;
            scorePlayerTwo += match.Winner == MatchWinner.PlayerTwo ? Bonus.Match : 0;

            return (scorePlayerOne, scorePlayerTwo);
        }

        private static (double, double) CalculateActualScore(int scoreTotalP1, int scoreTotalP2)
        {
            var actualScoreP1 = scoreTotalP1 > scoreTotalP2
                ? (1 - (scoreTotalP2 / (double)scoreTotalP1 / 2))
                : (scoreTotalP1 / (double)scoreTotalP2 / 2);

            return ((double)actualScoreP1, 1 - (double)actualScoreP1);
        }
    }
}

File: Frontend/frontend/src/app/_shared/engine/score.calculations.ts
import { BonusConstants as Bonus } from '../constants/bonus.constants';
import { Set } from '@entities/request/set';
import { MatchPlayer } from '@enums/match-player.enum';
import { MatchWinner } from '@enums/match-winner.enum';

export class ScoreCalculations {
  public static CalculateWinner(sets: Set[]): MatchWinner {
    let setsPlayerOne = 0;
    let setsPlayerTwo = 0;

    for (const set of sets) {
      if (!set) continue;
      if (set.gamesP1 && set.gamesP2) {
        setsPlayerOne += set.gamesP1 > set.gamesP2 ? 1 : 0;
        setsPlayerTwo += set.gamesP2 > set.gamesP1 ? 1 : 0;
      }
      if (set.pointsP1 && set.pointsP2) {
        setsPlayerOne +=
          set.isSuperTieBreaker() && set.pointsP1 > set.pointsP2 ? 1 : 0;
        setsPlayerTwo +=
          set.isSuperTieBreaker() && set.pointsP2 > set.pointsP2 ? 1 : 0;
      }
    }

    if (setsPlayerOne === setsPlayerTwo) return MatchWinner.Tie;
    return setsPlayerOne > setsPlayerTwo
      ? MatchWinner.PlayerOne
      : MatchWinner.PlayerTwo;
  }

  public static CalculateScoreTotals(
    sets: Set[],
    winner?: MatchPlayer
  ): [number, number] {
    let scorePlayerOne = 0;
    let scorePlayerTwo = 0;

    for (const set of sets) {
      if (set.gamesP1 && set.gamesP2) {
        scorePlayerOne += set.gamesP1 * Bonus.Game;
        scorePlayerTwo += set.gamesP2 * Bonus.Game;
        scorePlayerOne += set.gamesP1 > set.gamesP2 ? Bonus.Set : 0;
        scorePlayerTwo += set.gamesP2 > set.gamesP1 ? Bonus.Set : 0;
      }

      if (set.pointsP1 && set.pointsP2) {
        scorePlayerOne +=
          set.pointsP1 > set.pointsP2 && set.isTieBreaker()
            ? Bonus.TieBreaker
            : 0;
        scorePlayerTwo +=
          set.pointsP2 > set.pointsP2 && set.isTieBreaker()
            ? Bonus.TieBreaker
            : 0;
        scorePlayerOne +=
          set.pointsP1 > set.pointsP2 && set.isSuperTieBreaker()
            ? Bonus.SuperTieBreaker
            : 0;
        scorePlayerTwo +=
          set.pointsP2 > set.pointsP2 && set.isSuperTieBreaker()
            ? Bonus.SuperTieBreaker
            : 0;
      }
    }

    if (winner) {
      scorePlayerOne += winner === MatchPlayer.One ? Bonus.Match : 0;
      scorePlayerTwo += winner === MatchPlayer.Two ? Bonus.Match : 0;
    }

    return [scorePlayerOne, scorePlayerTwo];
  }

  public static GetWinnerText(
    providedWinner: MatchWinner | null | undefined,
    calcWinner: MatchWinner,
    opponentName: string
  ): string {
    const provided = providedWinner !== null;
    const agreement = providedWinner === calcWinner;

    const decap = (s: string) => s.charAt(0).toLowerCase() + s.slice(1);
    const tie = 'It was a tie';
    const him = () => `The winner is ${opponentName}`;
    const you = () => `You've won`;
    const looks = (result: string) => `It looks like ${decap(result)}`;
    const sure = (result: string, looksLike: string) =>
      `Are you sure ${decap(result)}? ${looksLike}`;

    if (!provided && calcWinner == null) {
      return 'There seems to be no winner';
    } else if (!provided && calcWinner === MatchWinner.Tie) {
      return looks(tie);
    } else if (!provided && calcWinner === MatchWinner.PlayerOne) {
      return looks(you());
    } else if (!provided) {
      return looks(him());
    } else if (agreement && providedWinner === MatchWinner.Tie) {
      return tie;
    } else if (agreement && providedWinner === MatchWinner.PlayerOne) {
      return `${you()} :)`;
    } else if (agreement) {
      return him();
    } else if (
      !agreement &&
      providedWinner === MatchWinner.Tie &&
      calcWinner === MatchWinner.PlayerOne
    ) {
      return sure(tie, looks(you()));
    } else if (
      !agreement &&
      providedWinner === MatchWinner.Tie &&
      calcWinner === MatchWinner.PlayerTwo
    ) {
      return sure(tie, looks(him()));
    } else if (
      !agreement &&
      providedWinner === MatchWinner.PlayerOne &&
      calcWinner === MatchWinner.PlayerTwo
    ) {
      return sure(you(), looks(him()));
    } else if (
      !agreement &&
      providedWinner === MatchWinner.PlayerOne &&
      calcWinner === MatchWinner.Tie
    ) {
      return sure(you(), looks(tie));
    } else if (
      !agreement &&
      providedWinner === MatchWinner.PlayerTwo &&
      calcWinner === MatchWinner.PlayerOne
    ) {
      return sure(him(), looks(you()));
    } else if (
      !agreement &&
      providedWinner === MatchWinner.PlayerTwo &&
      calcWinner === MatchWinner.Tie
    ) {
      return sure(him(), looks(tie));
    }

    return '';
  }
}
